website/types/user_activity.ts

```diff
@@ -1,12 +1,38 @@
+import type User from "#models/user"
+
+export enum UserActivityType {
+    Referral = 'referral',
+    AttendEvent = 'attend_event',
+    CompanyVisit = 'company_visit',
+    CompletedChallenge = 'completed_challenge'
+}
+
 export type UserActivityDescription = {
+    type: UserActivityType
+    description: ReferralDescription | AttendEventDescription | CompanyVisitDescription | CompletedChallengeDescription
+}
+    
+export type ReferralDescription = {
     referralCode: string
-    referralUserId: number // The user that has the referral link may not be a promoter (e.g. may not be a student associatio)
-    referredUserId: number
+    referralUser: User, // The user that has the referral link may not be a promoter (e.g. may not be a student associatio)
+    referredUser: User,
     // promoterId is used to determine if the referralUser is the promoter or only if the referral user was referred by the promoter 
     // thus giving the points to the promoter
     referralIsPromoter: boolean
     referralReferencedByPromoter?: boolean
     promoterId?: number 
     pointsToReferralUser?: number
     pointsToPromoter?: number
+}
+
+export type AttendEventDescription = {
+
+}
+
+export type CompanyVisitDescription = {
+
+}
+
+export type CompletedChallengeDescription = {
+
```

Comment:
This typing is suboptimal, we should have the type inside the description subtypes so typescript can handle type narrowing, something like:

```ts
type ReferralDescription = {
  type: 'referral'
  // Other stuff
  ...
}

type AttendEventDescription = {
  type: 'attend_event'
  // Other stuff
  ...
}

// Other types
...

type UserActivityDescription = ReferralDescription | AttendEventDescription | ...
```

This allows for the following code to have the correct auto complete:

```ts
if (description.type === 'referral')
  // description will now be of type ReferralDescription
```

If we still want `UserActivityType`, we can do `UserActivityDescription['type']`.